### PR TITLE
Update global checkpoint when increasing primary term on replica

### PR DESCRIPTION
**Pull request by ywelsch**
When a replica shard increases its primary term under the mandate of a new primary, it should also update its global checkpoint; this gives us the guarantee that its global checkpoint is at least as high as the new primary and gives a starting point for the primary/replica resync.

Relates to #25355, #10708

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -2072,6 +2084,13 @@ public void onResponse(final Releasable releasable) {
                                     primaryTerm);
                             onPermitAcquired.onFailure(new IllegalStateException(message));
                         } else {
+                            try {
+                                updateGlobalCheckpointOnReplica(globalCheckpoint);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Is it worth skipping this update if the update on term increase was successful?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 2]
```diff
@@ -354,7 +354,7 @@ public String toString() {
     private void ensureRefCount() {
         if (refCount() <= 0) {
             throw new ElasticsearchException("RecoveryStatus is used but it's refcount is 0. Probably a mismatch between incRef/decRef " +
-                    "calls");
+                "calls");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: All changes in this file look cosmetic, can they be reverted?

**Review by jasontedor — COMMENTED** [review]
I left a question and one comment.

**Comment by dengm129** [comment]
$25422

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: sure

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I found it easier to just do the "double"-update than introducing a boolean. Do you have a preference?

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by ywelsch** [comment]
Thanks @jasontedor

**Pull request merged**